numpy/lib/_npyio_impl.py

```diff
@@ -459,7 +459,7 @@ def load(file, mmap_mode=None, allow_pickle=False, fix_imports=True,
                 raise ValueError("Cannot load file containing pickled data "
                                  "when allow_pickle=False")
             try:
-                return pickle.load(fid, **pickle_kwargs)
+                return format.NumpyUnpickler(fid, **pickle_kwargs).load()
```

Comment:
Should we also wrap the call to `pickle.dump` to replace `_core` with `core` so numpy 2.x doesn't write `npy` files that are unreadable by numpy 1.x. Leave a comment that in a few years when downstream (hopefully) no longer cares about supporting NumPy 1.x we can remove that.

I don't think `NumpyUnpickler` should do this, it should only be done in the numpy file format internals to make extra sure files written in the for at are as compatible as possible.

Reply:
I think that each version (1.x and 2.0) should make their own pickles (with `core` and `_core` respectively). And user can employ `NumpyUnpickler` to read cross-version pickles.

If we replace `pickle.dump` in functions in `npyio` then it would mean that `np.save` creates numpy 1.x readable files but direct call to `pickle.dumps(arr)` won't - I think it would be confusing. I think that if someone is operating in numpy 2.0 realm only, they should be able to create pickles any possible way (with `np.save` and `pickle`) and read them without `NumpyUnpickler`.

Reply:
You convinced me. It would be weird for numpy to generate two different kinds of pickle data depending on how you do it. Point withdrawn.